### PR TITLE
Makes regen less greedy

### LINKED ISSUE #4298: group.skew() removes all tasks if group input is generator
## Checklist

- [X] I have included the output of ``celery -A proj report`` in the issue.
      (if you are not able to do this, then at least specify the Celery
       version affected).

      python 3.6.2
      celery==4.1.0
      amqp==2.2.1
      billiard==3.5.0.3

- [X] I have verified that the issue exists against the `master` branch of Celery.
      As of `be55de6`

## Steps to reproduce
```
@celery.task(bind=True)
def noop(task):
    pass
```

1. `group = celery.group(noop.si() for i in range(10))`
2. `group.skew()`  -> `group(<empty>)`
3. `group.apply_async()` or `group()`  -> `<GroupResult: 66fd517e-123b-4299-8cbe-9183b3f02626 []>`

## Expected behavior
```
<GroupResult: d610c923-e939-4199-9bb7-7bc89daa2ccb [9152bd60-f815-4bbe-a407-a003db34b19d, 7fc2d0ac-72ab-48b1-9083-be0f3b50b00d, d7c0a00a-85c3-4780-88f9-23be8da6fb73, 3737944f-91a4-4886-95c6-4960ef4764a7, a8da96ea-d2c4-4876-8026-e0c24c7d508a, 83ee5316-fb2f-472b-8e94-284144f92438, 5715a751-f28d-4deb-ac15-114854570fde, bc1646bf-cec9-4740-8e74-411d7b66dcca, 337aa1f3-5606-45b2-af68-5c5527289019, f5fffb98-f2f5-4423-a656-8c1d7e93291a]>
```

## Actual behavior

```
<GroupResult: 66fd517e-123b-4299-8cbe-9183b3f02626 []>
```

### Notes
If you `print(group)`(or something that evaluates `self.tasks`) before calling `group.skew()`, it works normally.

### PR DESCRIPTION
**Pull request by mjhoffman65**
Might fix #4298.  This was originally part of https://github.com/celery/celery/pull/6576 (https://github.com/celery/celery/pull/6576#pullrequestreview-561320820).  Related to https://github.com/celery/celery/pull/6542

### CODE REVIEW

**Comment by maybe-sybr** [comment]
@mjhoffman65 - apologies for the delay in getting #6542 in so you didn't need to worry about it. Just checking in quickly, can you tell me if this is significantly different from #6542 in any way other than rolling in the feedback from @auvipy about `__nonzero__()` vs `__bool__()`? I'll try to get either this or #6542 merged in the next day or so.

**Comment by auvipy** [comment]
> @mjhoffman65 - apologies for the delay in getting #6542 in so you didn't need to worry about it. Just checking in quickly, can you tell me if this is significantly different from #6542 in any way other than rolling in the feedback from @auvipy about `__nonzero__()` vs `__bool__()`? I'll try to get either this or #6542 merged in the next day or so.

you can push on this pr branch too. both of you get credit O:)

**Comment by mjhoffman65** [comment]
> > @mjhoffman65 - apologies for the delay in getting #6542 in so you didn't need to worry about it. Just checking in quickly, can you tell me if this is significantly different from #6542 in any way other than rolling in the feedback from @auvipy about `__nonzero__()` vs `__bool__()`? I'll try to get either this or #6542 merged in the next day or so.
> 
> you can push on this pr branch too. both of you get credit O:)

@maybe-sybr @auvipy I don't think this is different than #6542 other than the `__bool__` change.  Sounds good, if there's anything I can do let me know.  Feel free to push on this branch too if you go with this one.  No worries either way :-)

**Review by auvipy — APPROVED** [review]

**Comment by maybe-sybr** [comment]
In which case I'll just note that we didn't change to using a `yield from` noted on the other PR but that's not particularly important. Good to see this get in, thanks for resubbing it @mjhoffman65 and for the merge @auvipy

**Comment by thedrow** [comment]
FYI, `__nonzero__` is for Python 2.
We need to use `__bool__`, always.

> In which case I'll just note that we didn't change to using a `yield from` noted on the other PR but that's not particularly important. 

I missed that, I'll make another PR for it.

File: t/unit/utils/test_functional.py [hunk 1]
```diff
@@ -141,6 +141,7 @@ def build_generator():
         self.consumed_second_item = False
         g = regen(build_generator())
         assert bool(g)
+        assert g[0] == 1
```
Comment (thedrow) [inline t/unit/utils/test_functional.py:144]: Thanks for the test. I wasn't sure this was working correctly :)

**Pull request merged**